README.md
# hotel-reservation-clj

# Problem Statement

## Hotel Reservation Problem

A hotel chain operating in Miami wishes to offer room reservation services over the internet. They have three hotels: Lakewood, Bridgewood
and Ridgewood. Each hotel has separate weekday and weekend(Saturday and Sunday) rates. There are special rates for rewards customers as a
part of loyalty programs. Each hotel has a rating assigned to it.

* __Lakewood__
  * __Rating__: 3
  * __Weekday__ rates:
    * 110 for Regular;
    * 80 for Rewards.
  * __Weekend__ rates:
    * 90 for Regular;
    * 80 for Rewards.

* __Bridgewood__
  * __Rating__: 4
  * __Weekday__ rates:
    * 160 for Regular
    * 110 for Rewards.
  * __Weekend__ rates:
    * 60 for Regular;
    * 50 for Rewards.

* Ridgewood
  * __Rating:__ 5
  * __Weekday__ rates:
    * 220 for Regular;
    * 100 for Rewards.
  * __Weekend__ rates:
    * 150 for Regular;
    * 40 for Rewards.

Write a program to help an online customer find the cheapest hotel.

The input to the program will be a range of dates for a regular or rewards customer. The
output should be the cheapest available hotel. In case of a tie, the hotel with the highest
rating should be returned.

INPUT FORMAT:

```
<customer_type>: <date1>, <date2>, <date3>, ...
```

OUTPUT FORMAT:

```
<name_of_the_cheapest_hotel>
```

Examples:

INPUT 1:

```
Regular: 16Mar2009(mon), 17Mar2009(tues), 18Mar2009(wed)
```

OUTPUT 1:

```
Lakewood
```

INPUT 2:

```
Regular: 20Mar2009(fri), 21Mar2009(sat), 22Mar2009(sun)
```
OUTPUT 2:

```
Bridgewood
```

INPUT 3:

```
Rewards: 26Mar2009(thur), 27Mar2009(fri), 28Mar2009(sat)
```

OUTPUT 3:

```
Ridgewood
```

## Usage

    $ echo <input> | java -jar hotel-reservation-clj-0.1.0-standalone.jar


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-08-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-08-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hotel-reservation-clj/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hotel-reservation-clj/compare/0.1.0...0.1.1


test/hotel_reservation_clj/core_test.clj
(ns hotel-reservation-clj.core-test
  (:require [clojure.test :refer :all]
            [hotel-reservation-clj.core :as hotel]))

(deftest temp
  (testing "TBD"
    (is (= 1 1))))


test/hotel_reservation_clj/hotel_rating_test.clj
(ns hotel-reservation-clj.hotel-rating-test
  (:require [clojure.test :refer :all]
            [hotel-reservation-clj.hotel-rating :as rating]))

(deftest price-a-booking
  (def hotels [{:name "Lakewood"
                :rating 3
                :rates {:regular {:weekend 90 :weekday 110}
                        :rewards {:weekend 80 :weekday 80}}}
               {:name "Bridgewood"
                :rating 4
                :rates {:regular {:weekend 60 :weekday 160}
                        :rewards {:weekend 50 :weekday 110}}}
               {:name "Ridgewood"
                :rating 5
                :rates {:regular {:weekend 150 :weekday 220}
                        :rewards {:weekend 40  :weekday 100}}}])

  (is (= {:hotel "Lakewood" :rating 3 :price 330}
         (->>
          {:tier :regular :stay {:weekends 0 :weekdays 3}}
          (rating/price hotels))))

  (is (= {:hotel "Bridgewood" :rating 4 :price 280}
         (->>
          {:tier :regular :stay {:weekends 2 :weekdays 1}}
          (rating/price hotels))))

  (is (= {:hotel "Ridgewood" :rating 5 :price 240}
         (->>
          {:tier :rewards :stay {:weekends 1 :weekdays 2}}
          (rating/price hotels)))))


src/hotel_reservation_clj/hotel_rating.clj
(ns hotel-reservation-clj.hotel-rating
  (:gen-class))

(defn- price-or-rating
  [hotel-price-a hotel-price-b]
  (if (= (hotel-price-a :price) (hotel-price-b :price))
    (> (hotel-price-a :rating) (hotel-price-b :rating))
    (< (hotel-price-a :price) (hotel-price-b :price))))

(defn- get-rate
  [day-type hotel booking]
  (get-in hotel [:rates (booking :tier) day-type]))

(def ^:private weekend-rate (partial get-rate :weekend))
(def ^:private weekday-rate (partial get-rate :weekday))

(defn- to-priced-hotel
  [booking]
  (fn [hotel] {:hotel (hotel :name)
               :rating (hotel :rating)
               :price (+
                       (* (get-in booking [:stay :weekends])
                          (weekend-rate hotel booking))
                       (* (get-in booking [:stay :weekdays])
                          (weekday-rate hotel booking)))}))

(defn price
  [hotels booking]
  (->> hotels
       (map (to-priced-hotel booking))
       (into (sorted-set-by price-or-rating))
       first))



src/hotel_reservation_clj/booking_string.clj
(ns hotel-reservation-clj.booking-string
  (:gen-class)
  (:import [java.time LocalDate]
           [java.time DayOfWeek]
           [java.time.format DateTimeFormatter DateTimeParseException])
  (:require [clojure.string :as str]))

(def ^:private tiers
  {"Regular" :regular
   "Rewards" :rewards})

(def ^:private weekend-days #{DayOfWeek/SATURDAY DayOfWeek/SUNDAY})

(def ^:private date-format
  (DateTimeFormatter/ofPattern "ddMMMyyyy"))

(defn- weekend?
  [local-date]
  (contains? weekend-days (.getDayOfWeek local-date)))

(defn- error?
  [result]
  (not (nil? (result :error))))

(defn- error
  [message]
  {:error message})

(defn success
  [value]
  {:ok value})

(defn- invalid-date-error
  [date]
  (error (format "Invalid date: '%s'" date)))

(defn- invalid-customer-tier-error
  [tier]
  (error (format "Invalid customer tier: '%s'" tier)))

(defn remove-day-of-week
  [date-string]
  (subs date-string 0 9))

(defn- parse-customer-tier
  [customer-tier]
  (let [tier (tiers customer-tier)]
    (if (nil? tier)
      (invalid-customer-tier-error customer-tier)
      (success tier))))

(defn- date-string-to-local-date
  [date-string]
  (let [trimmed-date-string (str/trim date-string)]
    (try
      (-> trimmed-date-string
          remove-day-of-week
          (LocalDate/parse date-format)
          success)
      (catch Exception ex
        (invalid-date-error trimmed-date-string)))))

(defn- count-nights-of-stay
  [date-string]
  (defn process-date-by-date
    [dates stay]
    (if (empty? dates)
      (success stay)
      (let [date (-> dates
                     first
                     date-string-to-local-date)]
        (cond
          (error? date) date
          (weekend? (date :ok)) (process-date-by-date (rest dates) (update stay :weekends inc))
          :else (process-date-by-date (rest dates) (update stay :weekdays inc))))))

  (process-date-by-date (str/split date-string #",") {:weekends 0 :weekdays 0}))

(defn- parse-booking-string-chunks
  [tier-string dates-string]
  (let [parsed-tier (parse-customer-tier tier-string)
        nights-of-stay-count (count-nights-of-stay dates-string)]
    (cond
      (error? parsed-tier) parsed-tier
      (error? nights-of-stay-count) nights-of-stay-count
      :else (success {:tier (parsed-tier :ok)
                      :stay (nights-of-stay-count :ok)}))))

(defn- malformed-booking-string-error
  [booking-string]
  {:error (format "Malformed booking string '%s'" booking-string)})

(defn parse
  [booking-string]
  (let [[tier-chunk dates-chunk] (str/split booking-string #":")
        sanitised-tier-chunk (str/trim (or tier-chunk ""))
        sanitised-dates-chunk (str/trim (or dates-chunk ""))]
    (cond
      (empty? sanitised-tier-chunk) (malformed-booking-string-error booking-string)
      (empty? sanitised-dates-chunk) (malformed-booking-string-error booking-string)
      :else (parse-booking-string-chunks tier-chunk dates-chunk))))



src/hotel_reservation_clj/core.clj
(ns hotel-reservation-clj.core
  (:gen-class)
  (:require [hotel-reservation-clj.booking-string :as booking-string]
            [hotel-reservation-clj.hotel-rating :as rating]))

(def ^:private hotels [{:name "Lakewood"
                        :rating 3
                        :rates {:regular {:weekend 90 :weekday 110}
                                :rewards {:weekend 80 :weekday 80}}}
                       {:name "Bridgewood"
                        :rating 4
                        :rates {:regular {:weekend 60 :weekday 160}
                                :rewards {:weekend 50 :weekday 110}}}
                       {:name "Ridgewood"
                        :rating 5
                        :rates {:regular {:weekend 150 :weekday 220}
                                :rewards {:weekend 40  :weekday 100}}}])

(def ^:private to-priced-booking (partial rating/price hotels))

(defn to-best-hotel-or-error
  [booking]
  (condp = (-> booking keys first)
    :error (booking :error)
    :ok (-> booking
            :ok
            to-priced-booking
            :hotel)))

(defn -main
  []
  (->> (java.io.BufferedReader. *in*)
       line-seq
       (map booking-string/parse)
       (map to-best-hotel-or-error)
       println))
